api/python/tests/integration/test_packages.py

```diff
@@ -666,6 +672,70 @@ def test_s3_set_dir(self):
 
             list_object_versions_mock.assert_called_with('bucket', 'foo/')
 
+    def test_set_dir_update_policy(self):
+        """Verify building a package with update policy. """
+
+        # Test for local src dir
+        path = pathlib.Path(__file__).parent
+        data_dir = pathlib.Path(path, "data")
+        pkg = Package()
+        pkg = pkg.set_dir("/", data_dir, meta={'name': 'test_meta'})
+        assert 'nested' in pkg.keys()
+        assert pkg.meta == {'name': 'test_meta'}
+        existing_one_entry = pkg['nested/one.txt']
+
+        nested_dir = pathlib.Path(data_dir, "nested2")
+        nested_dir.mkdir(parents=True, exist_ok=True)
+        with open(nested_dir / 'one.txt', 'w') as fd:
+            fd.write(fd.name)
+
+        # existing update policy
+        pkg = pkg.set_dir("nested", nested_dir, update_policy='existing')
+        assert 'nested' in pkg.keys()
+        assert existing_one_entry == pkg['nested/one.txt']
+
+        # incoming update policy
+        pkg = pkg.set_dir("nested", nested_dir)
+        assert 'nested' in pkg.keys()
+        assert existing_one_entry != pkg['nested/one.txt']
+
+        # verify non existing update policy raises value error
+        expected_err = f"Update policy should be one of {str(PACKAGE_UPDATE_POLICY)}, not 'invalid_policy'"
+        with pytest.raises(ValueError) as e:
+            pkg.set_dir("nested", data_dir, update_policy='invalid_policy')
+        assert expected_err in str(e.value)
+
+        # remove created directory
+        shutil.rmtree(nested_dir)
+
+        # Test for s3 src dir
+        with patch('quilt3.packages.list_object_versions') as list_object_versions_mock:
+            list_object_versions_mock.return_value = ([
+                dict(Key='foo/a.txt', VersionId='xyz', IsLatest=True, Size=10),
+                dict(Key='foo/x/y.txt', VersionId='null', IsLatest=True, Size=10),
+                dict(Key='foo/z.txt', VersionId='123', IsLatest=False, Size=10),
+            ], [])
+            pkg = Package()
+            pkg.set_dir('', 's3://bucket/foo/', meta={'name': 'test_meta'})
+            assert pkg['a.txt'].get() == 's3://bucket/foo/a.txt?versionId=xyz'
+            assert pkg['x/y.txt'].get() == 's3://bucket/foo/x/y.txt?versionId=null'
+
+            list_object_versions_mock.return_value = ([
+                dict(Key='bar/a.txt', VersionId='abc', IsLatest=True, Size=10),
+                dict(Key='bar/x/y.txt', VersionId='null', IsLatest=True, Size=10),
+                dict(Key='bar/z.txt', VersionId='123', IsLatest=False, Size=10),
+            ], [])
+
+            # existing update policy
```

Comment:
once you parameterize you won't have this duplicated code. same for existing/incoming blocks above. whenever we have very similar code like that parameterize let's us consolidate the logic—very easy to maintain and extend as well.